README.md
# Dokumentace k 2. projektu do IPK 2023

Ondřej Zobal (xzobal01)

* * *

## Teorie

### TCP
    
 TCP je spolehlivý protokol transportní vrstvy, který pro rozlišení  cílové
aplikace používá šestnáctibitová čísla zvaná porty. Když skenujeme porty,
snažíme se zjistit, zda na konkrétní portu poslouchá nějaká  aplikace (a
případně jaká), leží ladem, nebo zda je komunikace tohoto  portu filtrována
firewallem.

 Než může aplikace začít komunikovat pomocí protokolu TCP musí nejprv  navázat
spojení. Prvně pošle žadatel paket SYN, kterým žádá o navázání  nového spojení,
adresát pak může s připojením souhlasit a poslat paket SYN ACK, nebo odmítnout a
poslat paket RST. Pokud adresát spjení přijal žadatel obdržel paket SYN ACK,
pošle vzdálenému počítači paket ACK, čímž se TCP spojení oficiálně považuje za
otevřené.

 Pro účely skenování můžeme tedy cílovému zařízení poslat segment SYN a  čekat,
jak odpoví. Pokud nazpátek obdržíme segment SYN ACK, znamená to, že je tento
port aktivně používán nějakou aplikací a je tudíž otevřený.  Pokud obdržíme
paket RST jako odpověď, znamená to, že tento port  nepřijímá žádná spojení a je
tedy zavřený. Pokud neobdržíme žádnou  odpověď, znamená to, že naši žádost
zablokoval paketový filtr.


### UDP

 UDP je téže protokol transportní vrstvy, avšak na rozdíl od TCP je  bezstavový
a neručí za bezpečné doručení zprávy. Podobně jako TCP i UDP používá
šestnáctibitové identifikátory pro rozlišení aplikací. Při  komunikaci přes UDP
posíláme data rovnou, bez nutnosti navazovat  spojení. Kvůli tomu nemůže
spolehlivě zjistit, zda je daný port  otevřený nebo filtrovaný. Jediným
indikátorem pro nás může být, že při  kontaktu se zavřeným portem některé
zařízení posílají chybové hlášení  protokolu ICMP, což neznačuje že daný port je


## Popis implementace

Tento program byl vytvořen pro operační systém GNU/Linux, na jiných operačních
systémech nemusí fungovat!

### Reprezentace portů v paměti

 Aby se porty na výstupu objevily ve stejném pořadí jako na vstupu,  použil jsem
TRP `PortMap`, která jako klíč používá číselnou hodnotu portu a jako data jeho
stav (tj. otevřený, zavřený, filtrovaný.) V kombinaci s  vektorem `PortEnumer`,
kde index reprezentuje pořadí portu na vstupu a  užitečnou hodnotou je pár
tvořený typem protokolu a číslem portu. Ve  vektoru pro výčet portů se tedy
objevují porty protokolu TCP i UDP,  zatímco TRP používám dvě, aby měl každý
protokol vlastní.


### Skenování

 Skenování probíhá souběžně ve třech vláknech. Hlavní vlákno programu  rozesílá
skenovací pakety a dvě vedlejší vlákna naslouchají příchozím  zprávám protokolů
TCP a ICMP. Obě přijímací vlákna zkoumají surové pakety svých protokolů  a
informace o portech zanáší do TRP.  Aby nedošlo k přehlcení koncového  zařízení,
mezi odesláním každého skenovacího paketu program počká 20  mikrosekund. 

 K výrobě i čtení paketů používám vestavěné struktury `iphdr`, `tcphdr`,
`udphdr` a `icmphdr`. Paměť obsahující paket přetypuji na příslušnou  strukturu
a s daty pracuji jako s jakoukoliv jinou strukturou jazyka C.  Jelikož používám
syrové sokety, musím kontrolovat, že příchozí segmenty  byly skutečně adresovány
mému skeneru.

 V moment kdy hlavní vlákno dokončí rozesílání všech paketů, pošle  signál, (v
podobě změny hodnoty atomické proměnné,) dvěma přijímacím  vláknům, která si od
toho okamžiku začnou odpočítávat časový interval specifikovaný na vstupu. Po
uplynutí časomíry obě vedlejší vlákna skončí.

### Tisk

 Tištění výstupu provádím postupným procházením portů ve vektoru portů a
vyhledáváním jejich stavu v TRP.


## Testování

 Program jsem testoval na svém serveru se zapnutým i vypnutým firewallem. Při
testování libovolného počtu TCP portů, výsledky přesně odpovídaly konfiguraci
zařízení. Při testování UDP portů, pouze prvních 6 testovaných zavřených portů
vrátilo chybový ICMP segment, tento limit je dán nastavením cílového zažízení

* * *


## Použité zdroje

> RFC 1071 - Computin the internet Checksum
> https://www.rfc-editor.org/rfc/rfc1071

> RFC 9293 - Transmission Control Protocol (TCP)
> https://www.rfc-editor.org/rfc/rfc9293

> RFC 768 - User Datagram Protocol
> https://www.rfc-editor.org/rfc/rfc768.html

> C++ reference 
> https://en.cppreference.com/w/

> NMAP - TCP SYN (Stealth) Scan (-sS)
> https://nmap.org/book/synscan.html

> OpenSource For U - A guide to using raw sockets
> https://www.opensourceforu.com/2015/03/a-guide-to-using-raw-sockets/

> Manuálové stránky jazyka C


CHANGELOG.md
# Changelog

Program funguje pouze na IPv4.

Program funguje pouze na Linuxu.


util.cc
#include <arpa/inet.h>
#include <netinet/ip.h>
#include <netinet/udp.h>
#include <netinet/ip_icmp.h>
#include <netinet/in.h>
#include <stdlib.h>
#include <cstring>
#include <iostream>
#include <net/if.h>

#include "util.hh"

using namespace std;

char* create_tcp_syn(const char* source, const char* dest, short dest_port, short source_port, int* packet_size) {
    *packet_size = sizeof(struct iphdr) + sizeof(struct tcphdr);
    char* buffer = static_cast<char*>(malloc(*packet_size));
    memset(buffer, 0, *packet_size);

    struct iphdr* ip_header = (struct iphdr*)(buffer);
    struct tcphdr* tcp_header = (struct tcphdr*)(buffer + sizeof(struct iphdr));

    ip_header->ihl = 5;
    ip_header->version = 4;
    ip_header->tos = 0;
    ip_header->tot_len = *packet_size;
    ip_header->id = rand() % 0xffff;
    ip_header->frag_off = 0;
    ip_header->ttl = 64;
    ip_header->protocol = IPPROTO_TCP;
    ip_header->saddr = inet_addr(source);
    ip_header->daddr = inet_addr(dest);
    ip_header->check = htons(htons(~(tcp_header->check)) + htons(~(ip_header->protocol)) + htons(sizeof(tcp_header) + sizeof(ip_header)));

    tcp_header->source = htons(source_port);
    tcp_header->dest = htons(dest_port);
    tcp_header->seq = rand();
    tcp_header->doff = 5;
    tcp_header->syn = 1;
    tcp_header->window = htons(8192);
    tcp_header->check = 0;
    tcp_header->urg_ptr = 0;

    struct checkSumIpHeader checkHeader;
    checkHeader.saddr = ip_header->saddr;
    checkHeader.daddr = ip_header->daddr;
    checkHeader.zero = '\0';
    checkHeader.protocol = ip_header->protocol;
    checkHeader.tcp_len = htons(sizeof(struct tcphdr));
    memcpy(&(checkHeader.tcp), tcp_header, sizeof(struct tcphdr));

    tcp_header->check = calculate_tcp_checksum((unsigned short*) &checkHeader, sizeof(struct checkSumIpHeader));

    return buffer;
}

char* create_udp_probe(const char* source, const char* dest, short dest_port, short source_port, int* packet_size) {
    *packet_size = sizeof(struct iphdr) + sizeof(struct udphdr);
    char* buffer = static_cast<char*>(malloc(*packet_size));
    memset(buffer, 0, *packet_size);

    struct iphdr* ip_header = (struct iphdr*)(buffer);
    struct udphdr* udp_header = (struct udphdr*)(((char*)buffer) + sizeof(struct iphdr));

    ip_header->ihl = 5;
    ip_header->version = 4;
    ip_header->tos = 0;
    ip_header->tot_len = *packet_size;
    ip_header->id = rand() % 0xffff;
    ip_header->frag_off = 0;
    ip_header->ttl = 64;
    ip_header->protocol = IPPROTO_UDP;
    ip_header->saddr = inet_addr(source);
    ip_header->daddr = inet_addr(dest);
    ip_header->check = htons(htons(~(udp_header->check)) + htons(~(ip_header->protocol)) + htons(sizeof(udp_header) + sizeof(ip_header)));

    udp_header->source = htons(source_port);
    udp_header->dest = htons(dest_port);
    udp_header->len = htons(sizeof(struct udphdr));
    udp_header->check = 0;

    return buffer;
}

int create_socket(int protocol, char* interface) {
    int raw_socket = socket(AF_INET, SOCK_RAW, protocol);
    if (raw_socket == -1) {
        cerr << "Could not create raw socket: " << strerror(errno) << endl;
        exit(1);
    }
    int on = 1;
    if (setsockopt(raw_socket, IPPROTO_IP, IP_HDRINCL, &on, sizeof(on)) == -1) {
        cerr << "Could not set IP_HDRINCL: " << strerror(errno) << endl;
        exit(1);
    }

    struct ifreq ifr;
    memset(&ifr, 0, sizeof(ifr));
    snprintf(ifr.ifr_name, sizeof(ifr.ifr_name), interface);
    if (setsockopt(raw_socket, SOL_SOCKET, SO_BINDTODEVICE, (void *)&ifr, sizeof(ifr)) < 0){
        cerr << "Could not bind interface" << endl;
        exit(1);
    }

    return raw_socket;
}

struct sockaddr_in create_target(char* destination) {
    struct sockaddr_in target;
    memset(&target, 0, sizeof(target));
    target.sin_family = AF_INET;
    target.sin_addr.s_addr = inet_addr(destination);
    return target;
}

// The following function was taken from RFC 1071 [https://www.rfc-editor.org/rfc/rfc1071]
// and is not subject to the GPL.
unsigned short calculate_tcp_checksum(unsigned short *addr, int count) {
    /* Compute Internet Checksum for "count" bytes
    *         beginning at location "addr".
    */
    long sum = 0;

    while( count > 1 )  {
        /*  This is the inner loop */
        sum += *addr++;
        count -= 2;
    }

    /*  Add left-over byte, if any */
    if( count > 0 )
        sum += * (unsigned char *) addr;

    /*  Fold 32-bit sum to 16 bits */
    while (sum>>16)
        sum = (sum & 0xffff) + (sum >> 16);

    return ~sum;
}


cmdline.cc
#include <sys/ioctl.h>
#include <net/if.h>
#include <sys/types.h>
#include <ifaddrs.h>
#include <ifaddrs.h>
#include <sys/socket.h>
#include <netdb.h>
#include <iostream>
#include <arpa/inet.h>
#include <netinet/ip.h>
#include <netinet/tcp.h>
#include <netinet/udp.h>
#include <netinet/ip_icmp.h>
#include <cstring>

#include "cmdline.hh"

using namespace std;

void list_interfaces() {
    struct ifaddrs *interfaces, *first;
    char addr_str[INET_ADDRSTRLEN];

    if (getifaddrs(&first) == -1) {
        cerr << "Canouln't list network interfaces" << endl;
        exit(1);
    }
    interfaces = first;

    while (interfaces != NULL) {
        if (interfaces->ifa_addr == NULL || interfaces->ifa_addr->sa_family != AF_INET) {
            interfaces = interfaces->ifa_next;
            continue;
        }

        struct sockaddr_in* sa = (struct sockaddr_in *)interfaces->ifa_addr;
        if (inet_ntop(AF_INET, &(sa->sin_addr), addr_str, INET_ADDRSTRLEN) == NULL) {
            cerr << "Couldn't find address of " << interfaces->ifa_name <<  endl;
            exit(1);
        }

        cout << interfaces->ifa_name << "\t" << addr_str << endl;
        interfaces = interfaces->ifa_next;
    }

    freeifaddrs(first);
}

void get_source_ip(const char* interface, string& ip) {
    struct ifaddrs *interfaces, *first;
    char addr_str[INET_ADDRSTRLEN];

    if (getifaddrs(&first) == -1) {
        cerr << "Canouln't list network interfaces" << endl;
        exit(1);
    }
    interfaces = first;

    while (interfaces != NULL) {
        if (interfaces->ifa_addr == NULL || interfaces->ifa_addr->sa_family != AF_INET) {
            interfaces = interfaces->ifa_next;
            continue;
        }

        if (strcmp(interfaces->ifa_name, interface)) {
            interfaces = interfaces->ifa_next;
            continue;
        }

        struct sockaddr_in* sa = (struct sockaddr_in *)interfaces->ifa_addr;
        if (inet_ntop(AF_INET, &(sa->sin_addr), addr_str, INET_ADDRSTRLEN) == NULL) {
            cerr << "Couldn't find address of " << interfaces->ifa_name <<  endl;
            exit(1);
        }
        ip = string(addr_str);
        break;
    }

    freeifaddrs(first);
}

string host_to_ip(string hostname) {
    struct addrinfo hints, *res;
    char ipstr[INET_ADDRSTRLEN];

    memset(&hints, 0, sizeof hints);
    hints.ai_family = AF_INET;
    hints.ai_socktype = SOCK_STREAM;

    if (getaddrinfo(hostname.c_str(), NULL, &hints, &res)) {
        cerr << "Couldn't convert " << hostname << " to IPv4 address" << endl;
        exit(1);
    }

    struct sockaddr_in* sock_addr = (struct sockaddr_in*)res->ai_addr;
    void* addr = &(sock_addr->sin_addr);
    inet_ntop(res->ai_family, addr, ipstr, sizeof ipstr);

    freeaddrinfo(res);

    return string(ipstr);
}

void parse_range(char* range, PortType type, PortEnumer& port_enumeration, PortMap& port_map) {
    string acc = "";

    bool trailing_separator;
    bool is_range = false;
    bool is_list = false;
    int bound;

    PortStatus status = (type == PortType::tcp) ? PortStatus::silent : PortStatus::open;

    for (char c = *range; c != '\0'; c = *(++range)) {
        trailing_separator = false;
        if ('0' <= c && c <= '9') {
            acc += c;
        }
        // List of ports ex. '10,20,22'
        else if (c == ',') {
            is_list = true;
            if (is_range) {
                cout << "Cannot combine range and list"  << endl;
                exit(1);
            }
            trailing_separator = true;

            int port = stoi(acc);
            acc = "";

            port_enumeration.push_back(pair(type, port));
            port_map.insert({port, status});

        }
        // Range ex. '10-100'
        else if (c == '-') {
            trailing_separator = true;
            // If '-' was specified twice.
            if (is_range || is_list) {
                cerr << "Invalid syntax for port range" << endl;
                exit(1);
            }

            is_range = true;
            bound = stoi(acc);
            acc = "";
        }
        else {
            cerr << c << endl;
            cerr << "Invalid port" << endl;
            exit(1);
        }
    }


    if (trailing_separator) {
        cerr << "Invalid syntax port" << endl;
        exit(1);
    }

    // Inserting ports in specified range
    if (is_range) {
        int other_bound = stoi(acc);
        if (bound >= other_bound) {
            cerr << "Invalid range" << endl;
            exit(1);
        }

        for (int i = bound; i <= other_bound; i++) {
            port_enumeration.push_back(pair(type, i));
            port_map.insert({i, status});
        }
    }
    // Inserting last port from list
    else {
        int port = stoi(acc);
        acc = "";

        port_enumeration.push_back(pair(type, port));
        port_map.insert({port, status});
    }
}

void process_cmdline_args(char** argv,
                          string& hostname,
                          int& limit_ms,
                          string& interface,
                          PortEnumer& port_enumeration,
                          PortMap& tcp_port_map,
                          PortMap& udp_port_map) {
    string prog_name = *argv;
    hostname = "";
    interface = "";

    while (*(++argv) != nullptr) {
        if (!strcmp(*argv, "-i") || !strcmp(*argv, "--interface")) {
            if (*(++argv) == nullptr) {
                list_interfaces();
                exit(1);
            }
            interface = string(*argv);
        }
        else if (!strcmp(*argv, "-t") || !strcmp(*argv, "--pt")) {
            if (*(++argv) == nullptr) {
                cerr << "Missing argument!" << endl;
                exit(1);
            }
            try {
                parse_range(*argv, PortType::tcp, port_enumeration, tcp_port_map);
            }
            catch (std::invalid_argument&) {
                cerr << "Bad port" << endl;
                exit(1);
            }
        }
        else if (!strcmp(*argv, "-u") || !strcmp(*argv, "--pu")) {
            if (*(++argv) == nullptr) {
                cerr << "Missing argument!" << endl;
                exit(1);
            }
            try {
                parse_range(*argv, PortType::udp, port_enumeration, udp_port_map);
            }
            catch (std::invalid_argument&) {
                cerr << "Bad port" << endl;
                exit(1);
            }
        }
        else if (!strcmp(*argv, "-w") || !strcmp(*argv, "--wait")) {
            if (*(++argv) == nullptr) {
                cerr << "Missing argument!" << endl;
                exit(1);
            }
            limit_ms = stoi(*argv) * 1000;
            if (limit_ms < 0) {
                cerr << "Can specify negative limits" << endl;
                exit(1);
            }
        }
        else {
            if (hostname != "") {
                cerr << "Hostname already specified" << endl;
                exit(1);
            }

            hostname = string(*argv);
        }
    }

    if (interface == "") {
        list_interfaces();
        exit(1);
    }

    if (hostname == "") {
        cerr << "No hostname given" << endl;
        exit(1);
    }
}


scanner.cc
#include <iostream>
#include <cstring>
#include <arpa/inet.h>
#include <netinet/ip.h>
#include <netinet/tcp.h>
#include <netinet/udp.h>
#include <netinet/ip_icmp.h>
#include <stdlib.h>
#include <unistd.h>
#include <signal.h>
#include <string>
#include <thread>
#include <chrono>
#include <time.h>
#include <mutex>
#include <atomic>
#include <sys/ioctl.h>
#include <net/if.h>

#include "scanner.hh"
#include "cmdline.hh"
#include "util.hh"


using namespace std;

/**
 * Sends TCP packet using a raw socket. Source and Destination addreses are specified in the numbers-and-dots notation.
 * */
void send_tcp_packet(char* source, char* destination, int d_port, int s_port, int raw_socket, struct sockaddr_in target) {
    int packet_size;
    char* tcp_packet = create_tcp_syn(source, destination, d_port, s_port, &packet_size);
    int ret;
    if ((ret = sendto(raw_socket, tcp_packet, packet_size, 0, (struct sockaddr *) &target, sizeof(target))) == -1) {
    }
    free(tcp_packet);
}

/**
 * Sends UDP packet using a raw socket. Source and Destination addreses are specified in the numbers-and-dots notation.
 * */
void send_udp_packet(char* source, char* destination, int d_port, int s_port, int raw_socket, struct sockaddr_in target) {
    int packet_size;
    char* udp_packet = create_udp_probe(source, destination, d_port, s_port, &packet_size);
    int ret;
    if ((ret = sendto(raw_socket, udp_packet, packet_size, 0, (struct sockaddr *) &target, sizeof(target))) == -1) {
    }
    free(udp_packet);
}

/**
 * Parses a TCP header, if it is relevant to scanning record it in port_map.
 * */
int process_tcp_response(struct tcphdr* tcp_header, unsigned short s_port, PortMap& port_map, std::mutex& port_map_mutex) {
    if (ntohs(tcp_header->dest) != s_port) {
        return 0;
    }

    PortStatus status;
    if (tcp_header->syn && tcp_header->ack) {
        status = PortStatus::open;
    } else {
        status = PortStatus::closed;
    }

    std::lock_guard<std::mutex> guard(port_map_mutex);
    if (port_map.find(ntohs(tcp_header->source)) != port_map.end()) {
        port_map[ntohs(tcp_header->source)] = status;
        return 1;
    }

    return 0;
}

/**
 * Parses an ICMP header, if it is relevant to scanning record it in port_map.
 * */
int process_icmp_response(struct icmphdr* icmp_header, unsigned short s_port, PortMap& udp_port_map, std::mutex& port_map_mutex) {
    if (icmp_header->type != ICMP_DEST_UNREACH || icmp_header->code != ICMP_PORT_UNREACH) {
        return 0;
    }

    // IP header of the original message is stored after 8 bytes
    struct iphdr* ip_header = (struct iphdr*) (((char*)icmp_header) + 8);
    struct udphdr* udp_header = (struct udphdr*) (((char*)ip_header) + sizeof(iphdr));

    if (ip_header->protocol != IPPROTO_UDP) {
        return 0;
    }

    if (udp_header->source != htons(s_port)) {
        return 0;
    }

    std::lock_guard<std::mutex> guard(port_map_mutex);
    if (udp_port_map.find(ntohs(udp_header->dest)) != udp_port_map.end()) {
        udp_port_map[ntohs(udp_header->dest)] = PortStatus::closed;
        return 1;
    }

    return 0;
}

/**
 * Listens for incoming packets and if they are a response to one of our probing segments,
 * records it's findings in port_map.
 * While sent_all is set to false, the function will receive indefinitely, once sent_all
 * turns true, the scanning will continue for limit_ms microseconds.
 * */
void recive_packet(int raw_socket,
                   unsigned short s_port,
                   PortMap& port_map,
                   std::mutex& port_map_mutex,
                   int limit_ms,
                   atomic<bool>* sent_all,
                   int protocol) {
    if (!port_map.size()) return;

    char response[1024 * 1024];
    memset(response, 0, sizeof(response));
    struct sockaddr_in sender;
    socklen_t sender_size = sizeof(sender);

    struct iphdr *ip_header = (struct iphdr *) response;
    char* next_header = (char *) (response + sizeof(struct iphdr));

    struct timeval timeout;
    chrono::time_point start = chrono::high_resolution_clock::now();
    int ms = -1;
    int recived = 0;

    do {
        // Doing this just so we can stop checking this atom once we se it as true once.
        if (ms == -1) {
            if (sent_all->load()) {
                start = chrono::high_resolution_clock::now();
                ms = 0;
            }
            else {
                // Timeout before we check again
                timeout.tv_sec = 0;
                timeout.tv_usec = 50;
            }
        }
        if (ms != -1) {
            // After we get a confirmation that the other thread sent all packets we can start the countdown on the
            // timeout.
            ms = chrono::duration_cast<std::chrono::milliseconds>(chrono::high_resolution_clock::now() - start).count();
            int remain = limit_ms - ms;

            if (remain <= 0 || (int) port_map.size() <= recived) {
                break;
            }

            timeout.tv_sec = remain / 1000; // convert to seconds
            timeout.tv_usec = (remain % 1000) * 1000; // convert remaining milliseconds to microseconds
        }

        setsockopt(raw_socket, SOL_SOCKET, SO_RCVTIMEO, &timeout, sizeof(timeout));

        int response_size = recvfrom(raw_socket, response, sizeof(response), 0, (struct sockaddr *) &sender, &sender_size);
        if (response_size == -1) continue;

        if (ip_header->protocol == IPPROTO_TCP && protocol == IPPROTO_TCP) {
            recived += process_tcp_response((struct tcphdr*) next_header, s_port, port_map, port_map_mutex);
        }
        else if (ip_header->protocol == IPPROTO_ICMP && protocol == IPPROTO_ICMP) {
            recived += process_icmp_response((struct icmphdr*) next_header, s_port, port_map, port_map_mutex);
        }

    } while (ms < limit_ms);
}

/**
 * Outputs information stored in TCP and UDP port maps to standard output.
 * */
void print_result(PortEnumer port_num, PortMap tcp_port_status, PortMap udp_port_status) {
    cout << "PORT STATE" << endl;
    for (pair i : port_num) {

        PortType type = i.first;
        unsigned short index = i.second;

        char* protocol;
        switch (type) {
            case PortType::tcp:
                protocol = (char*) "tcp";
                break;
            case PortType::udp:
                protocol = (char*) "udp";
                break;
        }

        char* status;
        if (type == PortType::tcp) {
            switch (tcp_port_status[index]) {
                case PortStatus::open:
                    status = (char*) "open";
                    break;
                case PortStatus::silent:
                    status = (char*) "filtered";
                    break;
                case PortStatus::closed:
                    status = (char*) "closed";
                    break;
            }
        }

        else if (type == PortType::udp) {
            switch(udp_port_status[index]) {
                case PortStatus::open:
                    status = (char*) "open";
                    break;
                case PortStatus::closed:
                    status = (char*) "closed";
                    break;
                default:;
            }
        }

        cout << i.second << "/" << protocol << " " << status << endl;
    }

}

/**
 * Sends one probing packet through the supplied raw socket for every entry in port_num.
 * Source and destination are specified using the numbers-and-dots notation.
 * */
void send_all_packets(char* source, char* destination, int s_port, int raw_socket, PortEnumer port_num) {
    struct sockaddr_in target = create_target(destination);

    for (auto port : port_num) {
        PortType type = port.first;
        unsigned short d_port = port.second;

        usleep(SEND_PAUSE_MS);

        if (type == PortType::tcp) {
            send_tcp_packet(source, destination, d_port, s_port, raw_socket, target);
        }
        else if (type == PortType::udp) {
            send_udp_packet(source, destination, d_port, s_port, raw_socket, target);
        }
    }
}

int main(int argc, char** argv) {
    (void) argc;
    srand(time(NULL));

    string source;
    string hostname;
    int limit_ms = 5000;
    string interface;

    PortEnumer port_num;
    PortMap tcp_port_status;
    PortMap udp_port_status;

    // Configuring program from command line arguments.
    process_cmdline_args(argv, hostname, limit_ms, interface, port_num, tcp_port_status, udp_port_status);

    // Converting supplied user-supplied hostname to an IP address.
    string destination = host_to_ip(hostname);
    // Picking a random 16-bit number for the port and making sure we don't colide
    // with any well-known ports.
    unsigned short s_port = rand() % (0xffff - 1024) + 1024;

    // Quaring address of selected network interface.
    get_source_ip((char*) interface.c_str(), source);

    // Creating raw sockets.
    // The TCP socket can send UDP packets just fine.
    int raw_socket = create_socket(IPPROTO_TCP, (char*) interface.c_str());
    int icmp_socket = create_socket(IPPROTO_ICMP, (char*) interface.c_str());

    // A flag used for communicating between threads that the sending of packets have been complete
    atomic<bool> sent_all { false };

    // Mutexes for locking PortMaps.
    std::mutex tcp_mutex;
    std::mutex udp_mutex;

    // The recvfrom function needs to run in it's ow thread because sending a large amounts of packets
    // takes a really long time and we could miss some responses.
    std::thread tcp_thread {recive_packet, raw_socket, s_port, ref(tcp_port_status), ref(tcp_mutex), limit_ms, &sent_all, IPPROTO_TCP};
    std::thread udp_thread {recive_packet, icmp_socket, s_port, ref(udp_port_status), ref(udp_mutex), limit_ms, &sent_all, IPPROTO_ICMP};

    // Sending probing packets.
    send_all_packets((char*) source.c_str(), (char*) destination.c_str(), s_port, raw_socket, port_num);
    sent_all.store(true);

    // Waiting for receiving threads.
    if (tcp_thread.joinable())
        tcp_thread.join();
    if (udp_thread.joinable())
        udp_thread.join();

    // We won't need the sockets anymore.
    close(raw_socket);
    close(icmp_socket);

    // Printing results.
    cout << "Interesting ports on " << hostname << " (" << destination << ")" << endl;
    print_result(port_num, tcp_port_status, udp_port_status);

    return 0;
}
